Source/WebCore/platform/network/curl/CurlMultipartHandle.cpp

```diff
@@ -34,309 +34,260 @@
+    const auto content = m_buffer.data();
+    auto boundaryLength = m_boundary.length();
+    const auto boundary = m_boundary.latin1().data();
 
-    boundaryStartPosition = notFound;
-    lastPartialMatchPosition = contentLength;
+    matchPos = notFound;
+    endPos = notFound;
+    hasCloseDelimiter = false;
 
-    if (contentLength < m_boundary.length())
-        return false;
+    for (size_t pos = 0; pos < contentLength; ++pos) {
+        size_t cmpLength = boundaryLength <= contentLength - pos ? boundaryLength : contentLength - pos;
 
-    const auto content = m_buffer.data();
+        if (memcmp(content + pos, boundary, cmpLength))
+            continue;
 
-    for (size_t i = 0; i < contentLength - m_boundary.length(); ++i) {
-        auto length = matchedLength(content + i);
-        if (length == m_boundary.length()) {
-            boundaryStartPosition = i;
+        auto afterBoundaryPos = pos + boundaryLength;
+
+        // The initial CRLF is considered to be attached to the boundary delimiter line rather than
+        // part of the preceding part. (See RFC2046 [5.1.1. Common Syntax])
+        if (pos - 2 >= 0 && !memcmp(content + pos - 2, "\r\n", 2))
+            pos -= 2;
+        else if (pos - 1 >= 0 && !memcmp(content + pos - 1, "\n", 1))
+            pos--;
+
+        // Not enough data to check.
+        if (cmpLength != boundaryLength) {
+            matchPos = pos;
+            return false;
+        }
+
+        if (afterBoundaryPos + 2 <= contentLength && !memcmp(content + afterBoundaryPos, "--", 2)) {
+            matchPos = pos;
+            endPos = afterBoundaryPos + 2;
+            hasCloseDelimiter = true;
             return true;
         }
 
-        if (length)
-            lastPartialMatchPosition = i;
-    }
+        // Skip transport-padding
+        for (; afterBoundaryPos < contentLength && isTabOrSpace(content[afterBoundaryPos]); ++afterBoundaryPos) { }
+
+        // There should be a \r and a \n but it seems that's not the case.
+        // So we'll check for a simple \n. Not really RFC compatible but servers do tricky things.
+        if (afterBoundaryPos + 2 <= contentLength && !memcmp(content + afterBoundaryPos, "\r\n", 2))
+            afterBoundaryPos += 2;
+        else if (afterBoundaryPos + 1 <= contentLength && !memcmp(content + afterBoundaryPos, "\n", 1))
+            afterBoundaryPos++;
+        else {
+            // Not enough data to check.
+            if (afterBoundaryPos >= contentLength) {
+                matchPos = pos;
+                return false;
+            }
 
-    return false;
-}
+            // A syntax error has occurred.
+            return std::nullopt;
+        }
 
-size_t CurlMultipartHandle::matchedLength(const uint8_t* data)
-{
-    auto length = m_boundary.length();
-    for (size_t i = 0; i < length; ++i) {
-        if (data[i] != m_boundary[i])
-            return i;
+        matchPos = pos;
+        endPos = afterBoundaryPos;
+        return true;
```

Comment:
If m_buffer contains "--delimiter" and not completed yet, and the next data will be "--", this code can't detect hasCloseDelimiter in such a case.

Reply:
Fixed

Reply:
Is this problem really fixed?

Reply:
I think I was able to fix the problem by following the steps below, but am I missing something?

https://github.com/WebKit/WebKit/pull/17403/files#diff-9748a6c95db83f4a038a66c392226087d69df3953308a3f96c081618d378e750R239
```
    if (needMoreData || (!m_didCompleteMessage && contentEndPtr - matchedEndPtr < 2)) {
        // Not enough data to check the boundary or the Close Delimiter.
```

Reply:
Oh, sorry. I misread.